File: src/client/java/fr/hardel/asset_editor/client/javafx/lib/utils/ColorUtils.java
package fr.hardel.asset_editor.client.javafx.lib.utils;

import javafx.scene.paint.Color;

public final class ColorUtils {

    public static int stringToHue(String text) {
        int hash = 0;
        for (int i = 0; i < text.length(); i++) {
            hash = text.charAt(i) + ((hash << 5) - hash);
        }
        return (int) (Math.abs((long) hash) % 360L);
    }

    public static Color hueToColor(int hue) {
        double h = ((hue % 360) + 360) % 360;
        double s = 0.70;
        double l = 0.60;
        double c = (1 - Math.abs(2 * l - 1)) * s;
        double x = c * (1 - Math.abs((h / 60.0) % 2 - 1));
        double m = l - c / 2.0;
        double r = 0;
        double g = 0;
        double b = 0;
        if (h < 60) {
            r = c;
            g = x;
        } else if (h < 120) {
            r = x;
            g = c;
        } else if (h < 180) {
            g = c;
            b = x;
        } else if (h < 240) {
            g = x;
            b = c;
        } else if (h < 300) {
            r = x;
            b = c;
        } else {
            r = c;
            b = x;
        }
        return Color.color(clamp(r + m), clamp(g + m), clamp(b + m));
    }

    public static String toCssRgba(Color color) {
        return "rgba(%d,%d,%d,%.3f)".formatted(
                (int) Math.round(color.getRed() * 255),
                (int) Math.round(color.getGreen() * 255),
                (int) Math.round(color.getBlue() * 255),
                color.getOpacity());
    }

    private static double clamp(double value) {
        return Math.max(0.0, Math.min(1.0, value));
    }

    private ColorUtils() {
    }
}

File: src/client/kotlin/fr/hardel/asset_editor/client/compose/lib/utils/ColorUtils.kt
package fr.hardel.asset_editor.client.compose.lib.utils

import androidx.compose.ui.graphics.Color
import kotlin.math.abs
import kotlin.text.iterator

object ColorUtils {

    fun accentColor(key: String?): Color {
        if (key.isNullOrBlank()) return hueToColor(0)
        return hueToColor(stringToHue(key.trim()))
    }

    fun stringToHue(text: String): Int {
        var hash = 0
        for (c in text) {
            hash = c.code + ((hash shl 5) - hash)
        }
        return (abs(hash.toLong()) % 360L).toInt()
    }

    fun hueToColor(hue: Int, saturation: Float = 0.70f, lightness: Float = 0.60f): Color {
        val h = ((hue % 360) + 360) % 360
        val s = saturation.coerceIn(0f, 1f)
        val l = lightness.coerceIn(0f, 1f)
        val c = (1f - abs(2f * l - 1f)) * s
        val x = c * (1f - abs((h / 60f) % 2f - 1f))
        val m = l - c / 2f

        val (r, g, b) = when {
            h < 60 -> Triple(c, x, 0f)
            h < 120 -> Triple(x, c, 0f)
            h < 180 -> Triple(0f, c, x)
            h < 240 -> Triple(0f, x, c)
            h < 300 -> Triple(x, 0f, c)
            else -> Triple(c, 0f, x)
        }

        return Color(
            red = (r + m).coerceIn(0f, 1f),
            green = (g + m).coerceIn(0f, 1f),
            blue = (b + m).coerceIn(0f, 1f)
        )
    }
}